benchmarks/functional_autograd_benchmark/functional_autograd_benchmark.py

```diff
@@ -0,0 +1,155 @@
+from typing import NamedTuple, Callable, List, Any
+
+import ppl_models
+import vision_models
+import audio_text_models
+
+from utils import to_markdown_table, TimingResultType, InpType, GetterType, VType
+
+# Listing of the different tasks
+FAST_TASKS_NO_DOUBLE_BACK = [
+    "vjp",
+]
+
+FAST_TASKS = FAST_TASKS_NO_DOUBLE_BACK + [
+    "vhp",
+    "jvp",
+]
+
+ALL_TASKS = FAST_TASKS + [
+    "hvp",
+    "jacobian",
+    "hessian"
+]
+
+DOUBLE_BACKWARD_TASKS = ["jvp", "hvp", "vhp", "hessian"]
+
+# Model definition which contains:
+# - name: a string with the model name.
+# - getter: a function to get the model. It takes as input the device on which the model
+#     will run. It should return the forward function and the parameters (Tensors) used as
+#     input for the forward function. Note that the forward must *not* have any side effect.
+# - tasks: the list of recommended tasks that can run in a reasonable amount of time with this model.
+# - unsupported: the list of tasks that this model cannot run.
+class ModelDef(NamedTuple):
+    name: str
+    getter: GetterType
+    tasks: List[str]
+    unsupported: List[str]
+
+MODELS = [
+    ModelDef("resnet18", vision_models.get_resnet18, FAST_TASKS, []),
+    ModelDef("fcn_resnet", vision_models.get_fcn_resnet, FAST_TASKS, []),
+    ModelDef("detr", vision_models.get_detr, FAST_TASKS, []),
+    ModelDef("ppl_simple_reg", ppl_models.get_simple_regression, ALL_TASKS, []),
+    ModelDef("ppl_robust_reg", ppl_models.get_robust_regression, ALL_TASKS, []),
+    ModelDef("wav2letter", audio_text_models.get_wav2letter, FAST_TASKS, []),
+    ModelDef("deepspeech", audio_text_models.get_deepspeech, FAST_TASKS_NO_DOUBLE_BACK, DOUBLE_BACKWARD_TASKS),
+    ModelDef("transformer", audio_text_models.get_transformer, FAST_TASKS, []),
+    ModelDef("multiheadattn", audio_text_models.get_multiheadattn, FAST_TASKS, []),
+]
+
+def get_v_for(model: Callable, inp: InpType, task: str) -> VType:
+    v: VType
+
+    if task in ["vjp"]:
+        out = model(*inp)
+        v = torch.rand_like(out)
+    elif task in ["jvp", "hvp", "vhp"]:
+        if isinstance(inp, tuple):
+            v = tuple(torch.rand_like(i) for i in inp)
+        else:
+            v = torch.rand_like(inp)
+    else:
+        v = None
+
+    return v
+
+def run_once(model: Callable, inp: InpType, task: str, v: VType) -> None:
+    func = getattr(functional, task)
+
+    if v is not None:
+        res = func(model, inp, v=v, strict=True)
+    else:
+        res = func(model, inp, strict=True)
+
+def run_model(model_getter: GetterType, args: Any, task: str) -> List[float]:
+    if args.gpu == -1:
+        device = torch.device("cpu")
+
+        def noop():
+            pass
+        do_sync = noop
+    else:
+        device = torch.device("cuda:{}".format(args.gpu))
+        do_sync = torch.cuda.synchronize
+
+    model, inp = model_getter(device)
+
+    v = get_v_for(model, inp, task)
+    # Warmup
+    run_once(model, inp, task, v)
```

Comment:
nit: for some benchmarking applications, folks will run a few (more than 1) warmup iterations and have that be controllable by command line. I'm not sure how much that affects the models that are being run here

Reply:
I think most of the models here are fairly large. So the warmup is mostly for the cuda caching allocator to allocate stuff and the models to do caching if they do so.
Maybe we can add this later if we need to.